src/features/payment/PaymentProvider.tsx

```diff
@@ -0,0 +1,80 @@
+import React, { createContext, useContext, useEffect, useRef, useState } from 'react';
+import type { ReactNode } from 'react';
+
+import type { AxiosError } from 'axios';
+
+import { Loader } from 'src/core/loading/Loader';
+import { useProcessNavigation } from 'src/features/instance/ProcessNavigationContext';
+import { usePaymentInformation } from 'src/features/payment/PaymentInformationProvider';
+import { PaymentStatus } from 'src/features/payment/types';
+import { usePerformPayActionMutation } from 'src/features/payment/usePerformPaymentMutation';
+import { useIsPayment } from 'src/features/payment/utils';
+import { useNavigationParam } from 'src/features/routing/AppRoutingContext';
+
+type PaymentContextProps = {
+  setLoading: (bool) => void;
+  performPayment: () => void;
+  paymentError: AxiosError | null;
+};
+
+type PaymentContextProvider = {
+  children: ReactNode;
+};
+
+export const PaymentContext = createContext<PaymentContextProps | undefined>(undefined);
+
+export const PaymentProvider: React.FC<PaymentContextProvider> = ({ children }) => {
+  const [loading, setLoading] = useState<boolean>(false);
+  const partyId = useNavigationParam('partyId');
+  const instanceGuid = useNavigationParam('instanceGuid');
+  const paymentInfo = usePaymentInformation();
+  const { next, busy } = useProcessNavigation() || {};
+  const { mutate, error } = usePerformPayActionMutation(partyId, instanceGuid);
+
+  const contextValue: PaymentContextProps = {
+    setLoading,
+    performPayment: () => {
+      setLoading(true);
+      mutate();
+    },
+    paymentError: error,
+  };
+
+  const paymentDoesNotExist = paymentInfo?.status === PaymentStatus.Uninitialized;
+  const isPaymentProcess = useIsPayment();
+  const actionCalled = useRef(false);
+
+  // if PaymentStatus is Uninitialized, initiate it by calling the pay action and go to payment provider
+  useEffect(() => {
+    if (isPaymentProcess && paymentDoesNotExist && !actionCalled.current) {
+      actionCalled.current = true;
+      setLoading(true);
+      mutate();
+    }
+  }, [isPaymentProcess, paymentDoesNotExist, mutate]);
+
+  const paymentCompleted = paymentInfo?.status === PaymentStatus.Paid || paymentInfo?.status === PaymentStatus.Skipped;
+  const nextCalled = useRef(false);
+
+  useEffect(() => {
+    if (paymentCompleted && next && !busy && !nextCalled.current) {
+      nextCalled.current = true;
+      setLoading(true);
+      next({ action: 'confirm', nodeId: 'next-button' });
+    }
+  }, [paymentCompleted, next, busy]);
+
+  if (loading) {
+    return <Loader reason='Navigating to external payment solution' />;
+  }
+
+  return <PaymentContext.Provider value={contextValue}>{children}</PaymentContext.Provider>;
```

Comment:
I suspect this can lead to re-renders of all form components when anything in this provider changes. The `contextValue` object is at least not memoized, so this context will provide a new value every re-render. Have you checked if this affects anything when (for example) submitting a form (not necessarily a form with payment)? If useProcessNavigation() causes this to re-render because process navigation became busy, that might lead to expensive re-renders immediately when the submit button is pressed.

Reply:
Btw, these things are difficult to get right, which is why we made `createZustandContext()` and used that pattern in many of these global form-providers. That way you can set all the state you want, change it however you want, but never cause full-tree re-renders unless every component in the tree uses the state from zustand.

Reply:
I'll do some testing and see if i can make use of the pattern with `createZustandContext` 😊 👍